Rebase `eip7805` fork onto Electra

jihoonsong opened the pull request:
0.
This PR rebases `eip7805` fork onto Electra. The rest two commits correct IL submission behavior according to the EIP (i.e., submission deadline is 8 second into the slot whereas propagation cut off is 9).

After this PR, we can have Prysm and Geth interop with the following config:
```
  electra_fork_epoch: 0
  eip7805_fork_epoch: 1
```

Current implementation of `ethereum-package` sets Osaka time only when Fulu fork is set. Our original Geth's `ForkchoiceUpdatedV3` checks if current fork version is Osaka so it doesn't work without Fulu. To make Prysm and Geth work, I've uploaded Geth image that doesn't check fork version within `ForkchoiceUpdatedV3`. It's available at `jihoonsg/geth-focil:no-osaka`.

To sum up, this [Kurtosis config](https://github.com/jihoonsong/local-devnet-focil/blob/main/kurtosis/kurtosis.config.registry.yaml) reflects these two changes.

If we want to set Osaka time, we can set fork epochs as:
```
  electra_fork_epoch: 0
  fulu_fork_epoch: 1
  eip7805_fork_epoch: 2
```

While this works like a charm, it invalidates the whole effort to rebase onto Electra.

One hacky thing allows us to have both fulu and eip7805 fork at epoch 1. At `UpgradeToEip7805` in `beacon-chain/core/eip7805/upgrade.go`, we could use `Fork.PreviousVersion: params.BeaconConfig().ElectraForkVersion` instead of `Fork.PreviousVersion: params.BeaconConfig().CurrentVersion`.

This is because when we upgrade states, fulu state changes occur right before eip7805 state changes within the same epoch. So we end up with `Fork.PreviousVersion == FuluForkVersion`, which should have been `Fork.PreviousVersion == ElectraForkVersion`. (You can refer to `UpgradeState` in `beacon-chain/core/transition.go`.)

With this change, this following config works:
```
  electra_fork_epoch: 0
  fulu_fork_epoch: 1
  eip7805_fork_epoch: 1
```

However, this is very error-prone and high-contextual workaround, which I don't really like. Needless to say, having fulu fork activated is the opposite of why we're trying to rebase onto Electra, i.e., to not affected by PeerDAS and other changes. Hence, I'd like to suggest to use eip7805 fork in the EL as well to become independent from fulu completely. I want to discuss this in the upcoming FOCIL breakout.

1.
I think we need fork.md something like [that of fulu](https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/fork.md#upgrading-the-state). I'll open a PR to consensus-spec adding fork.md soon.

2.
This PR is heavily inspired by #14771. I'd like to appreiciate the PR author and reviewer @nalepae @rkapka

terencechain committed:
Implement EIP7805: Fork-choice enforced Inclusion Lists

jihoonsong committed:
Emit IL event after successfully receive an IL over the P2P (#15033)

jihoonsong committed:
Calculate IL submission and view freeze deadlines relative to seconds_per_slot

jihoonsong committed:
Submit IL one second before submission deadline

jihoonsong commented:
A PR adding fork.md for eip7805: https://github.com/ethereum/consensus-specs/pull/4191

jihoonsong commented:
I found a bug that this change kills Dora after eip7805 fork activated epoch is passed. I'll add a fixing commit.

terencechain reviewed (changes requested):
Please note that the team is generally against boilerplate — for example, #14771 was later reverted in https://github.com/prysmaticlabs/prysm/pull/14905  
I don't think we should add any EIP-7805 related boilerplate and should only include the minimal code necessary to test functionality  
Could we strip out the parts we don't need in this PR (ex: block and state do not change in EIP7805) and review it based on that? Otherwise this PR becomes too big to review

ensi321 commented on beacon-chain/core/eip7805/upgrade.go:
```diff
@@ -0,0 +1,175 @@
+		NextWithdrawalValidatorIndex: vi,
+		HistoricalSummaries:          summaries,
+
+		DepositRequestsStartIndex:     params.BeaconConfig().UnsetDepositRequestsStartIndex,
```
`DepositRequestsStartIndex` just needs to be `beaconState.DepositRequestsStartIndex()`. Setting to `UnsetDepositRequestsStartIndex` is a one time operation when upgrading to Electra

jihoonsong committed:
Rebase `eip7805` fork onto `Electra`

jihoonsong commented:
I've removed boilerplates and reduced changed files by ~66%. There might be more changes that are not necessary like configuring minimal or Sepolia config, etc. I've removed things a bit conservatively so please let me know if those are better to be gone.

This changes also not kill Dora after the transition to epoch 2.

jihoonsong committed:
Add `DOMAIN_IL_COMMITTEE` preset

terencechain committed:
Update FOCIL to v6

jihoonsong committed:
Add eip7805 config to Hoodi config
